### PR TITLE
{lang}[GCCcore/10.3.0] Python v3.9.5

### PR DESCRIPTION
**Pull request by SebastianAchilles**
(created using `eb --new-pr`)

Depends on
- [x] `GMP/6.2.1` ~#12613~
- [x] ~`OpenSSL/1.1.1k` (optional) #12686~
- [x] `OpenSSL/1.1` ~#12858~
- [x] `Rust/1.52.1` ~#12904~

Not the latest version in the EC for Python v3.9.5 are:
- ~`docutils`, because `sphinx 3.5.4 has requirement docutils<0.17,>=0.12`,~
- `idna`, because `requests 2.25.1 has requirement idna<3,>=2.5`,
- `keyring`, because `poetry 1.1.6 has requirement keyring<22.0.0,>=21.2.0`,
- ~`cryptography 3.2.1`, which is the most recent `cryptography` with `OpenSSL` 1.0.x support.~

In addition:
- `cryptography 3.4.7` requires `Rust`, `setuptools-rust` and `semantic_version` ~as well as `CRYPTOGRAPHY_DONT_BUILD_RUST=1`~,
- ~`blist` is failing without `modulename: False`.~

CC: @boegel @Micket @robert-mijakovic

### CODE REVIEW

**Comment by SebastianAchilles** [comment]
Test report by @sebastianachilles
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
jrlogin11.jureca - Linux centos linux 8.3.2011, x86_64, AMD EPYC 7742 64-Core Processor, Python 3.6.8
See https://gist.github.com/6e07edc67c81224fc6884bf61a27a6fb for a full test report.

**Review by Micket — APPROVED** [review]
lgtm
I don't see a reason to hold back on this any longer?

**Comment by Micket** [comment]
@boegelbot please test @ generoso

**Comment by SebastianAchilles** [comment]
> I don't see a reason to hold back on this any longer?

Maybe we want to include the changes from https://github.com/easybuilders/easybuild-easyconfigs/pull/12650
But I am not sure about that. What is your opinion?

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
bear-pg0211u03a.bear.cluster - Linux RHEL 8.3, x86_64, Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz (cascadelake), Python 3.6.8
See https://gist.github.com/2ccd9b59a8299622f6ff63621bd6b372 for a full test report.

**Comment by SebastianAchilles** [comment]
Test report by @sebastianachilles
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
ebvm.novalocal - Linux centos linux 8.3.2011, x86_64, Intel Xeon Processor (Skylake, IBRS), Python 3.6.8
See https://gist.github.com/ce862523dc9e731c31008fc372dcfbe3 for a full test report.

**Comment by verdurin** [comment]
Test report by @verdurin
**SUCCESS**
Build succeeded for 7 out of 7 (1 easyconfigs in total)
nuc.lan - Linux Fedora 33, x86_64, Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz, Python 3.9.2
See https://gist.github.com/e21b76acb523b527a013dda3d114808a for a full test report.

**Comment by robert-mijakovic** [comment]
Should we bump it to 3.9.5?

File: easybuild/easyconfigs/p/Python/Python-3.9.4-GCCcore-10.3.0.eb [hunk 1]
```diff
@@ -0,0 +1,442 @@
+name = 'Python'
+version = '3.9.4'
```
Comment (boegel) [inline easybuild/easyconfigs/p/Python/Python-3.9.4-GCCcore-10.3.0.eb:2]: @SebastianAchilles We definitely want this bumped to 3.9.5...

File: easybuild/easyconfigs/p/Python/Python-3.9.4-GCCcore-10.3.0.eb [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+osdependencies = [OS_PKG_OPENSSL_DEV]
+
+# symlink 'pip' command to 'pip3' that is included with Python installation
+# required so we can update pip to version included in extensions, using pip
+installopts = " && ln -s %(installdir)s/bin/pip3 %(installdir)s/bin/pip"
+
+exts_default_options = {
+    'download_dep_fail': True,
+    'sanity_pip_check': True,
+    'source_urls': [PYPI_SOURCE],
```
Comment (boegel) [inline easybuild/easyconfigs/p/Python/Python-3.9.4-GCCcore-10.3.0.eb:44]: This should no longer be needed, since `PythonPackage` now automatically uses this by default.

File: easybuild/easyconfigs/p/Python/Python-3.9.4-GCCcore-10.3.0.eb [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+    'use_pip': True,
+}
+
+
```
Comment (boegel) [inline easybuild/easyconfigs/p/Python/Python-3.9.4-GCCcore-10.3.0.eb:48]: please remove duplicate empty line

**Review by boegel — CHANGES_REQUESTED** [review]

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 1 out of 2 (1 easyconfigs in total)
node3125.skitty.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 3.6.8
See https://gist.github.com/38f6deb6f8c317d183895be03a7f4c1c for a full test report.

**Comment by SebastianAchilles** [comment]
@boegel I have added `[PYPI_SOURCE]` as `source_urls` again. Unfortunately I didn't noticed the issue since I had the extensions in my `$EASYBUILD_SOURCEPATH`.

File: easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+    # it's nice to have an up to date openssl for security reasons
+    # Python 3.7 requires OpenSSL >= 1.0.2, uncomment line below if OS version is too old (e.g. CentOS 6)
+    # ('OpenSSL', '1.1.1k'),
+]
+
+osdependencies = [OS_PKG_OPENSSL_DEV]
+
+# symlink 'pip' command to 'pip3' that is included with Python installation
+# required so we can update pip to version included in extensions, using pip
+installopts = " && ln -s %(installdir)s/bin/pip3 %(installdir)s/bin/pip"
```
Comment (Flamefire) [inline easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb:39]: Note: Superflous after https://github.com/easybuilders/easybuild-easyblocks/pull/2388 is merged

File: easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+exts_default_options = {
+    'download_dep_fail': True,
+    'sanity_pip_check': True,
+    'source_urls': [PYPI_SOURCE],
+    'use_pip': True,
+}
+
+# order is important!
+# package versions updated 10 May 2021
+exts_list = [
```
Comment (Flamefire) [inline easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb:50]: Order must be: setuptools, wheel, pip due to dependence, see https://github.com/easybuilders/easybuild-easyconfigs/pull/12650

File: easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+
+# symlink 'pip' command to 'pip3' that is included with Python installation
+# required so we can update pip to version included in extensions, using pip
+installopts = " && ln -s %(installdir)s/bin/pip3 %(installdir)s/bin/pip"
+
+exts_default_options = {
+    'download_dep_fail': True,
+    'sanity_pip_check': True,
+    'source_urls': [PYPI_SOURCE],
+    'use_pip': True,
```
Comment (Flamefire) [inline easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb:45]: Until https://github.com/easybuilders/easybuild-easyblocks/pull/2388 lands, please add `'pip_ignore_installed': False`

File: easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+}
+
+# order is important!
+# package versions updated 10 May 2021
+exts_list = [
+    ('pip', '21.1.1', {
+        'use_pip': False,
```
Comment (Flamefire) [inline easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb:52]: You can/must use pip to upgrade pip:
```suggestion
```

Note: Needs the reordering

File: easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb [hunk 8]
```diff
@@ -0,0 +1,441 @@
...
+    }),
+    ('setuptools', '56.2.0', {
+        'source_tmpl': '%(name)s-%(version)s.tar.gz',
+        'checksums': ['7bb5652625e94e73b9358b7ed8c6431b732e80cf31f4e0972294c64f0e5b849e'],
+    }),
+    ('nose', '1.3.7', {
+        'checksums': ['f1bffef9cbc82628f6e7d7b40d7e255aefaa1adb6a1b1d26c69a8b79e6208a98'],
+    }),
+    ('blist', '1.3.6', {
+        'modulename': False,
```
Comment (Flamefire) [inline easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb:66]: > blist is failing without modulename: False.

How is it failing? It looks like something might be wrong then. Do you still have the error message?

File: easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb [hunk 9]
```diff
@@ -0,0 +1,441 @@
...
+    ('flit', '3.2.0', {
+        'checksums': ['592464c9268bbacec9bc67b5a3ae62e6e090aeec1563e69501df338a1728e551'],
+    }),
+    ('regex', '2021.4.4', {
+        'checksums': ['52ba3d3f9b942c49d7e4bc105bb28551c44065f139a65062ab7912bef10c9afb'],
+    }),
+    ('intreehooks', '1.0', {
+        'checksums': ['87e600d3b16b97ed219c078681260639e77ef5a17c0e0dbdd5a302f99b4e34e1'],
+    }),
+    ('crashtest', '0.3.1', {
```
Comment (Flamefire) [inline easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb:341]: Move this below poetry and remove the `use_pip` line

File: easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb [hunk 10]
```diff
@@ -0,0 +1,441 @@
...
+    }),
+    ('pastel', '0.2.1', {
+        'use_pip': False,
+        'checksums': ['e6581ac04e973cac858828c6202c1e1e81fee1dc7de7683f3e1ffe0bfd8a573d'],
+    }),
+    ('clikit', '0.6.2', {
+        'use_pip': False,
+        'checksums': ['442ee5db9a14120635c5990bcdbfe7c03ada5898291f0c802f77be71569ded59'],
+    }),
+    ('jeepney', '0.6.0', {
```
Comment (Flamefire) [inline easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb:356]: Move this below poetry and remove the `use_pip` line

File: easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb [hunk 11]
```diff
@@ -0,0 +1,441 @@
...
+        'checksums': ['87e600d3b16b97ed219c078681260639e77ef5a17c0e0dbdd5a302f99b4e34e1'],
+    }),
+    ('crashtest', '0.3.1', {
+        'use_pip': False,
+        'checksums': ['42ca7b6ce88b6c7433e2ce47ea884e91ec93104a4b754998be498a8e6c3d37dd'],
+    }),
+    ('pylev', '1.3.0', {
+        'checksums': ['063910098161199b81e453025653ec53556c1be7165a9b7c50be2f4d57eae1c3'],
+    }),
+    ('pastel', '0.2.1', {
```
Comment (Flamefire) [inline easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb:348]: Can you use the whl and pip: https://github.com/easybuilders/easybuild-easyconfigs/pull/12650/files#diff-62174d1a70710afde8442e36d7bd56cb93bf98a80eaa010139cd3770c5f9df49R337

File: easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb [hunk 12]
```diff
@@ -0,0 +1,441 @@
...
+        'use_pip': False,
+        'checksums': ['e6581ac04e973cac858828c6202c1e1e81fee1dc7de7683f3e1ffe0bfd8a573d'],
+    }),
+    ('clikit', '0.6.2', {
```
Comment (Flamefire) [inline easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb:352]: Same here: using the whl file works (and is what `pip install clikit` would do)

File: easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb [hunk 13]
```diff
@@ -0,0 +1,441 @@
...
+    ('keyring', '21.8.0', {
+        'modulename': False,  # Doesn't work properly if HOME directory contains keys
+        'checksums': ['1746d3ac913d449a090caf11e9e4af00e26c3f7f7e81027872192b2398b98675'],
+    }),
+    ('keyrings.alt', '4.0.2', {
+        'modulename': False,
+        'checksums': ['cc475635099d6edd7e475c5a479e5b4da5e811a3af04495a1e9ada488d16fe25'],
+    }),
+    ('tomlkit', '0.7.0', {
+        'use_pip': False,
```
Comment (Flamefire) [inline easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb:372]: Same: `'source_tmpl': SOURCE_WHL,`

File: easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb [hunk 14]
```diff
@@ -0,0 +1,441 @@
...
+        'checksums': ['75ecc79f26d99222a084fbdd1ce5aad3ac3a8bd535cd9059528452da38b68841'],
+    }),
+    ('webencodings', '0.5.1', {
+        'checksums': ['b36a1c245f2d304965eb4e0a82848379241dc04b865afcc4aab16748587e1923'],
+    }),
+    ('html5lib', '1.1', {
+        'checksums': ['b2e5b40261e20f354d198eae92afc10d750afb487ed5e50f9c4eaf07c184146f'],
+    }),
+    ('cleo', '0.8.1', {
+        'use_pip': False,
```
Comment (Flamefire) [inline easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb:403]: Same: `'source_tmpl': SOURCE_WHL,`

File: easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb [hunk 15]
```diff
@@ -0,0 +1,441 @@
...
+        'checksums': ['186581f4ceb42a0bbe040c407da73c14092379b1e4c0e327fdb72ae4a9b269b1'],
+    }),
+    ('msgpack', '1.0.2', {
+        'checksums': ['fae04496f5bc150eefad4e9571d1a76c55d021325dcd484ce45065ebbdd00984'],
+    }),
+    ('CacheControl', '0.12.6', {
+        'checksums': ['be9aa45477a134aee56c8fac518627e1154df063e85f67d4f83ce0ccc23688e8'],
+    }),
+    ('ptyprocess', '0.7.0', {
+        'use_pip': False,
```
Comment (Flamefire) [inline easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb:416]: Same: 'source_tmpl': SOURCE_WHL,

**Review by Flamefire — CHANGES_REQUESTED** [review]
As mentioned in the conf call here the required changes.

TLDR of most comments: Use pip to install all packages as I've seen issues when not doing so to differences how meta data is handled. See https://github.com/easybuilders/easybuild-easyconfigs/pull/12650

- 2 extensions couldn't be installed because they require poetry -> Move down
- The others have generic wheels which can be used with pip
- Reordering of setuptools, wheel, pip is required (wheel depends on setuptools, pip on both, setuptools on nothing)
- 1 change required until https://github.com/easybuilders/easybuild-easyblocks/pull/2388 is merged and 1 can be done after that is merged

File: easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb (hunk 5 above)
Comment (SebastianAchilles) [reply easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb:50]: If I use this order I am getting `error: invalid command 'bdist_wheel'` for `setuptools`. I assume the dependencies of `setuptools` changed. According to the documentation of `setuptools` v56.2.0 `wheel` is needed.

For me the order `wheel, setuptools, pip` worked.

File: easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb (hunk 8 above)
Comment (SebastianAchilles) [reply easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb:66]:
```
== 2021-05-12 12:09:34,583 build_log.py:169 ERROR EasyBuild crashed with an error (at easybuild/base/exceptions.py:124 in __init__): Sanity check failed: extensions sanity check failed for 1 extensions: blist
failing sanity check for 'blist' extension: command "python -c "import blist"" failed; output:
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/p/usersoftware/slai/jurecadc/stages/2021a/software/Python/3.9.5-GCCcore-10.3.0/lib/python3.9/site-packages/blist/__init__.py", line 2, in <module>
    from blist._blist import *
ImportError: /p/usersoftware/slai/jurecadc/stages/2021a/software/Python/3.9.5-GCCcore-10.3.0/lib/python3.9/site-packages/blist/_blist.cpython-39-x86_64-linux-gnu.so: undefined symbol: _PyObject_GC_IS_TRACKED,  (at easybuild/framework/easyblock.py:2872 in _sanity_check_step)
```
This is a bug of `blist` with Python >= 3.9:
https://bugzilla.redhat.com/show_bug.cgi?id=1830801
https://github.com/DanielStutzbach/blist/issues/90

Unfortunately the blist upstream repo has been inactive for years. Do we want to remove the package? Otherwise there is a patch for `blist` to work with Python >=3.9:
https://github.com/conda-forge/blist-feedstock/blob/master/recipe/patches/0004-compatibility-with-PEP-620.patch

File: easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb (hunk 9 above)
Comment (SebastianAchilles) [reply easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb]: Moved down.
I replaced `'use_pip': False,` with `'source_tmpl': SOURCE_PY3_WHL,` Otherwise I couldn't download the package.

**Comment by SebastianAchilles** [comment]
With https://github.com/easybuilders/easybuild-easyconfigs/pull/12755/commits/aa2faaee085187705687a820aef9fd0c001d19a3 I have addressed the requested changes from @Flamefire to consistently use pip to install Python packages.

I have also added a patch for `blist` to work with Python 3.9. But maybe we want to remove this package?

**Comment by SebastianAchilles** [comment]
Test report by @sebastianachilles
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
jrlogin11.jureca - Linux centos linux 8.3.2011, x86_64, AMD EPYC 7742 64-Core Processor, Python 3.6.8
See https://gist.github.com/4ddce792faeaca485037766ff8bacbae for a full test report.

File: easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb (hunk 5 above)
Comment (Flamefire) [reply easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb:51]: Wow, great. So wheel needs setuptools and setuptools needs wheel :D
But glad you figured this out, this works

**Comment by Flamefire** [comment]
Test report by @Flamefire
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
taurusi8014 - Linux centos linux 7.9.2009, x86_64, AMD EPYC 7352 24-Core Processor (zen2), Python 2.7.5
See https://gist.github.com/e68141b35f7725319369c5f05e3f575b for a full test report.

**Comment by Flamefire** [comment]
Test report by @Flamefire
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
taurusi8014 - Linux centos linux 7.9.2009, x86_64, AMD EPYC 7352 24-Core Processor (zen2), Python 2.7.5
See https://gist.github.com/898fc85986bd4b95949745e69efd335e for a full test report.

**Comment by boegel** [comment]
@Flamefire That's due to cryptography not supported OpenSSL 1.0.1 anymore, which is the version that comes with CentOS 7.x...

**Comment by Flamefire** [comment]
Test report by @Flamefire
**FAILED**
Build succeeded for 0 out of 27 (1 easyconfigs in total)
taurusi6214.taurus.hrsk.tu-dresden.de - Linux RHEL 7.9, x86_64, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz (haswell), Python 2.7.5
See https://gist.github.com/f6d6d24ed8fa25248d032546e24d4000 for a full test report.

**Comment by SebastianAchilles** [comment]
I downgraded `cryptography` to 3.1.1 which  is the most recent `cryptography` with `openssl` 1.0.x support.

**Comment by Flamefire** [comment]
> I downgraded cryptography to 3.1.1 which is the most recent cryptography with openssl 1.0.x support.

Can we go with 3.2.1? In 3.3 support for 1.0.2 was removed by https://github.com/pyca/cryptography/commit/d3eae8d7dbcd7ca491531424a4ac8b4838acf199 This would (likely) work on CENTOS 7

**Comment by boegel** [comment]
@boegelbot please test @ generoso

**Comment by branfosj** [comment]
> > I downgraded cryptography to 3.1.1 which is the most recent cryptography with openssl 1.0.x support.
> 
> Can we go with 3.2.1? In 3.3 support for 1.0.2 was removed by [pyca/cryptography@d3eae8d](https://github.com/pyca/cryptography/commit/d3eae8d7dbcd7ca491531424a4ac8b4838acf199) This would (likely) work on CENTOS 7

I went off the release notes in suggesting 3.1.1: https://github.com/pyca/cryptography/blob/main/CHANGELOG.rst#32---2020-10-25

It looks like an env variable needs setting to use 3.2.1: https://github.com/pyca/cryptography/blob/0c7592c34fd58f0634f493d6ce542ab35d940b26/src/cryptography/hazmat/bindings/openssl/binding.py#L174

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
node3153.skitty.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 3.6.8
See https://gist.github.com/3196a2edc3766ea1f2cc2e2a1d98060b for a full test report.

**Comment by boegel** [comment]
Indeed, the sanity check fails currently:
```
== 2021-05-12 18:20:54,413 run.py:226 INFO running cmd: python -c 'from cryptography.hazmat.bindings.openssl import binding'
== 2021-05-12 18:20:54,724 build_log.py:169 ERROR EasyBuild crashed with an error (at easybuild/easybuild-framework/easybuild/base/exceptions.py:124 in __init__): cmd "python -c 'from cryptography.hazmat.bindings.openssl import binding'" exited with exit code 1 and output:
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/user/gent/400/vsc40023/eb_arcaninescratch/CO7/skylake-ib/software/Python/3.9.5-GCCcore-10.3.0/lib/python3.9/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 222, in <module>
    _verify_openssl_version(Binding.lib)
  File "/user/gent/400/vsc40023/eb_arcaninescratch/CO7/skylake-ib/software/Python/3.9.5-GCCcore-10.3.0/lib/python3.9/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 182, in _verify_openssl_version
    raise RuntimeError(
RuntimeError: You are linking against OpenSSL 1.0.2, which is no longer supported by the OpenSSL project. To use this version of cryptography you need to upgrade to a newer version of OpenSSL. For this version only you can also set the environment variable CRYPTOGRAPHY_ALLOW_OPENSSL_102 to allow OpenSSL 1.0.2.
 (at easybuild/easybuild-framework/easybuild/tools/run.py:564 in parse_cmd_output)
```

Sticking to a previous version of cryptography is a relatively easy way out, but the question remains whether it's the *correct* way out. And we're bound to hit similar problems due to OpenSSL 1.0.1 in CentOS 7 with other software packages? (cfr. discussion in the EasyBuild Slack + #11895).

**Comment by Flamefire** [comment]
Test report by @Flamefire
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
taurusi8015 - Linux centos linux 7.9.2009, x86_64, AMD EPYC 7352 24-Core Processor (zen2), Python 2.7.5
See https://gist.github.com/538fb4c1f02f24ec066417249a25e5cb for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 6 out of 6 (1 easyconfigs in total)
bask-pg0309u05a - Linux RHEL 8.3, x86_64, Intel(R) Xeon(R) Platinum 8358P CPU @ 2.60GHz (icelake), Python 3.6.8
See https://gist.github.com/450f542f3c14a6afc1266dcb9bf15c92 for a full test report.

**Review by lexming — COMMENTED** [review]
Maybe we should move cryptography into its own easyconfig. This would allow the user to choose between:
* a cryptography easyconfig that depends on OpenSSL v1.1.x in EB (with the planed wrappers #11895) 
* a cryptography easyconfig using the system OpenSSL

Moreover, recent cryptography v3.4.7 depends on `setuptools-rust`, which would mean (at some point) adding rust as build dependency to Python. I think that this is unnecessary complexity for the base Python easyconfig.

I checked the current extensions in the Python easyconfig and only paramiko, SecretStorage, keyring and poetry need cryptography. So it's easy to split them out. We could even make a Python-crypto bundle with these extensions.

**Comment by boegel** [comment]
@lexming I like that idea, especially with the Rust build dependency aspect, we definitely don't want that for Python itself...

It's a good way out here, but not a general solution for the OpenSSL mess in CentOS 7.
But I'm not sure there's a good way out there anyway without causing trouble somewhere else.

**Review by boegel — CHANGES_REQUESTED** [review]
@SebastianAchilles With https://github.com/easybuilders/easybuild-easyblocks/pull/2388 and https://github.com/easybuilders/easybuild-easyconfigs/pull/12650 merged, a couple of things should be updated in here, incl. adding `install_pip = True`

**Comment by boegel** [comment]
`install_pip = True` set with last commit ()

I tried taking out the `cryptography` extension, but as a result the following had to be removed too (because they depend on `cryptography` directly or indirectly): `paramiko`, `SecretStorage`, `keyring`, `keyrings.alt`, `poetry`.

Especially the latter is painful I think, since `poetry` is a required build tool for getting several other Python packages properly installed.

There's a potential alternate way out via an OpenSSL wrapper proposed by @lexming, see https://github.com/easybuilders/easybuild-easyconfigs/issues/11895#issuecomment-841301807, we should try that approach first...

File: easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb [hunk 16]
```diff
@@ -0,0 +1,434 @@
...
+    ('XZ', '5.2.5'),
+    ('GMP', '6.2.1'),  # required for pycrypto
+    ('libffi', '3.3'),  # required for cryptography
+    # OS dependency should be preferred if the os version is more recent then this version,
+    # it's nice to have an up to date openssl for security reasons
+    # Python 3.7 requires OpenSSL >= 1.0.2, uncomment line below if OS version is too old (e.g. CentOS 6)
+    # ('OpenSSL', '1.1.1k'),
+]
+
+osdependencies = [OS_PKG_OPENSSL_DEV]
```
Comment (boegel) [inline easybuild/easyconfigs/p/Python/Python-3.9.5-GCCcore-10.3.0.eb:35]: @SebastianAchilles Please change to depend on the `OpenSSL/1.1` wrapper (added in #12858), see also #12922

**Review by boegel — CHANGES_REQUESTED** [review]

**Comment by boegel** [comment]
The bot was acting stupid a bit, by posting the same comment over and over again.

That should be a lot less common now thanks to https://github.com/boegel/boegelbot/commit/287e6e1359f7093a11ff3b58b4cf9b3e122557fd

**Comment by boegel** [comment]
@boegelbot please test @ generoso
CORE_CNT=16

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
select-pika-c6gd-2xlarge-0001 - Linux centos linux 8.3.2011, AArch64, ARM UNKNOWN (graviton2), Python 3.6.8
See https://gist.github.com/71c176e01ddd5d0af7cefb31a73f141d for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
node3515.doduo.os - Linux RHEL 8.2, x86_64, AMD EPYC 7552 48-Core Processor (zen2), Python 3.6.8
See https://gist.github.com/7ab20d1244f0c200996ae7a4bdf3d39d for a full test report.

**Comment by SebastianAchilles** [comment]
Test report by @sebastianachilles
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
jrlogin11.jureca - Linux centos linux 8.3.2011, x86_64, AMD EPYC 7742 64-Core Processor, Python 3.6.8
See https://gist.github.com/e62da7d666c5dc85d72eb4f21bfabe2c for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
easybuild2.novalocal - Linux centos linux 8.3.2011, POWER, IBM pSeries (emulated by qemu) (power9le), Python 3.6.8
See https://gist.github.com/65f651f4632a24d08bf30afcd7545778 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (1 easyconfigs in total)
node2618.swalot.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz (haswell), Python 3.6.8
See https://gist.github.com/e6ea37277074308a31baf49060acf6ca for a full test report.

**Review by boegel — APPROVED** [review]

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
node3129.skitty.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 3.6.8
See https://gist.github.com/e78d607b6c311e731e44c83381cebfc0 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
select-pika-c6gd-2xlarge-0001 - Linux centos linux 8.3.2011, AArch64, ARM UNKNOWN (graviton2), Python 3.6.8
See https://gist.github.com/0987598f5f79e48aaa4e58cbd50d518f for a full test report.

**Comment by boegel** [comment]
Going in, thanks @SebastianAchilles!
